lib/internal/modules/esm/loader.js

```diff
@@ -340,35 +349,58 @@ class ModuleLoader {
     //    evaluated at this point.
     // TODO(joyeecheung): add something similar to CJS loader's requireStack to help
     // debugging the the problematic links in the graph for import.
+    debug('importSyncForRequire', parent?.filename, '->', filename, job);
     if (job !== undefined) {
       mod[kRequiredModuleSymbol] = job.module;
       const parentFilename = urlToFilename(parent?.filename);
       // TODO(node:55782): this race may stop to happen when the ESM resolution and loading become synchronous.
+      let raceMessage = `Cannot require() ES Module ${filename} because it is not yet fully loaded. `;
+      raceMessage += 'This may be caused by a race condition if the module is simultaneously dynamically ';
+      raceMessage += 'import()-ed via Promise.all(). Try await-ing the import() sequentially in a loop instead.';
+      if (parentFilename) {
+        raceMessage += ` (from ${parentFilename})`;
+      }
       if (!job.module) {
-        let message = `Cannot require() ES Module ${filename} because it is not yet fully loaded. `;
-        message += 'This may be caused by a race condition if the module is simultaneously dynamically ';
-        message += 'import()-ed via Promise.all(). Try await-ing the import() sequentially in a loop instead.';
-        if (parentFilename) {
-          message += ` (from ${parentFilename})`;
-        }
-        assert(job.module, message);
+        assert(job.module, raceMessage);
```

Comment:
We might as well
```suggestion
        assert.fail(raceMessage);
```

Reply:
If we do care about the cost of generating the race message then the if block is necessary or otherwise it's unconditionally generated..